aeppl/loglik.py

```diff
@@ -0,0 +1,209 @@
+        ``var.tag.value_var`` will be checked and, when available, used.
+
+    """
+    if not isinstance(rv_values, Mapping):
+        rv_values = {var: rv_values} if rv_values is not None else {}
+
+    rv_var, rv_value_var = extract_rv_and_value_vars(var)
+
+    rv_value = rv_values.get(rv_var, rv_value_var)
+
+    if rv_var is not None and rv_value is None:
+        raise ValueError(f"No value variable specified or associated with {rv_var}")
+
+    if rv_value is not None:
+        rv_value = at.as_tensor(rv_value)
+
+        if rv_var is not None:
+            # Make sure that the value is compatible with the random variable
+            rv_value = rv_var.type.filter_variable(rv_value.astype(rv_var.dtype))
+
+        if rv_value_var is None:
+            rv_value_var = rv_value
+
+    if rv_var is None:
+        if var.owner is not None:
+            return _loglik(
+                var.owner.op,
+                var,
+                rv_values,
+                *var.owner.inputs,
+            )
+
+        return at.zeros_like(var)
+
+    rv_node = rv_var.owner
+
+    rng, size, dtype, *dist_params = rv_node.inputs
+
+    # Here, we plug the actual random variable into the log-likelihood graph,
+    # because we want a log-likelihood graph that only contains
+    # random variables.  This is important, because a random variable's
+    # parameters can contain random variables themselves.
+    # Ultimately, with a graph containing only random variables and
+    # "deterministics", we can simply replace all the random variables with
+    # their value variables and be done.
+
+    # tmp_rv_values = rv_values.copy()
+    # tmp_rv_values[rv_var] = rv_var
+
+    logpdf_var = _logpdf(rv_node.op, rv_value_var, *dist_params, **kwargs)
+
+    # Replace random variables with their value variables
+    replacements = rv_values.copy()
+    replacements.update({rv_var: rv_value, rv_value_var: rv_value})
+
+    (logpdf_var,), _ = rvs_to_value_vars(
+        (logpdf_var,),
+        initial_replacements=replacements,
+    )
+
+    if sum:
+        logpdf_var = at.sum(logpdf_var)
+
+    # Recompute test values for the changes introduced by the replacements
+    # above.
+    if config.compute_test_value != "off":
+        for node in io_toposort(graph_inputs((logpdf_var,)), (logpdf_var,)):
+            compute_test_value(node)
+
+    if rv_var.name is not None:
+        logpdf_var.name = "__logp_%s" % rv_var.name
+
+    return logpdf_var
+
+
+@singledispatch
+def _loglik(
+    op: Op,
+    var: TensorVariable,
+    rvs_to_values: Dict[TensorVariable, TensorVariable],
+    *inputs: TensorVariable,
+    **kwargs,
+):
+    """Create a graph for the log-likelihood of a ``Variable``.
+
+    This function dispatches on the type of ``op``.  If you want to implement
+    new graphs for an ``Op``, register a new function on this dispatcher.
+
+    The default returns a graph producing only zeros.
+
+    """
+    value_var = rvs_to_values.get(var, var)
```

Comment:
My gut feeling when I was working on this, is that this default is more error prone. I feel most Aesara ops in a generative model would have some impact on the total logp. Wouldn't it be safer to default to NotImplementedError and explicitly dispatch those that should indeed return zero?

Reply:
All this is going to be refactored.  We only need the `logpdf`s for `aehmc` right now.